halo2-base/src/safe_types/mod.rs

```diff
@@ -0,0 +1,136 @@
+/// do any extra value checking if they take SafeType as inputs.
+#[derive(Clone, Debug)]
+pub struct SafeType<F: ScalarField, const BYTES_PER_ELE: usize, const TOTAL_BITS: usize> 
+where [(); (TOTAL_BITS + BYTES_PER_ELE * BITS_PER_BYTE - 1) / (BYTES_PER_ELE * BITS_PER_BYTE)]: Sized {
+    // value is stored in little-endian. (BYTES_PER_ELE * BITS_PER_BYTE) is the number of bits of a single element.
+    value: [AssignedValue<F>; (TOTAL_BITS + BYTES_PER_ELE * BITS_PER_BYTE - 1) / (BYTES_PER_ELE * BITS_PER_BYTE)],
+}
+
+impl<F: ScalarField, const BYTES_PER_ELE: usize, const TOTAL_BITS: usize> SafeType<F, BYTES_PER_ELE, TOTAL_BITS> 
+where [(); (TOTAL_BITS + BYTES_PER_ELE * BITS_PER_BYTE - 1) / (BYTES_PER_ELE * BITS_PER_BYTE)]: Sized {
+    pub const BYTES_PER_ELE: usize = BYTES_PER_ELE;
+    pub const TOTAL_BITS: usize = TOTAL_BITS;
+    pub const BITS_PER_ELE: usize = min(TOTAL_BITS, BYTES_PER_ELE * BITS_PER_BYTE);
+    pub const VALUE_LENGTH: usize = (TOTAL_BITS + BYTES_PER_ELE * BITS_PER_BYTE - 1) / (BYTES_PER_ELE * BITS_PER_BYTE);
+
+    // new is private so Safetype can only be constructed by this crate.
+    fn new(raw_values: RawAssignedValues<F>) -> Self {
+        Self { value: raw_values.try_into().unwrap() }
+    }
+
+    // Return values in littile-endian.
+    pub fn value(&self) -> &[AssignedValue<F>; (TOTAL_BITS + BYTES_PER_ELE * BITS_PER_BYTE - 1) / (BYTES_PER_ELE * BITS_PER_BYTE)] {
+        &self.value
+    }
+}
+
+/// Represent TOTAL_BITS with the least number of AssignedValue<F>.
+/// (2^(F::NUM_BITS) - 1) might not be a valid value for F. e.g. max value of F is a prime in [2^(F::NUM_BITS-1), 2^(F::NUM_BITS) - 1]
+type CompactSafeType<F: ScalarField, const TOTAL_BITS: usize> = SafeType<F, { ((F::NUM_BITS - 1) / 8) as usize}, TOTAL_BITS>;
+
+/// SafeType for bool.
+pub type SafeBool<F> = CompactSafeType<F, 1>;
+/// SafeType for uint8.
+pub type SafeUint8<F> = CompactSafeType<F, 8>;
+/// SafeType for uint16.
+pub type SafeUint16<F> = CompactSafeType<F, 16>;
+/// SafeType for uint32.
+pub type SafeUint32<F> = CompactSafeType<F, 32>;
+/// SafeType for uint64.
+pub type SafeUint64<F> = CompactSafeType<F, 64>;
+/// SafeType for uint128.
+pub type SafeUint128<F> = CompactSafeType<F, 128>;
+/// SafeType for uint256.
+pub type SafeUint256<F> = CompactSafeType<F, 256>;
+/// SafeType for bytes32.
+pub type SafeBytes32<F> = SafeType<F, 1, 256>;
+
+pub struct SafeTypeChip<F: ScalarField> {
+    pub range_chip: Arc<RangeChip<F>>,
+}
+
+impl<F: ScalarField> SafeTypeChip< F> {
+    /// Construct a SafeTypeChip.
+    pub fn new(range_chip: Arc<RangeChip<F>>) -> Self {
+        Self { range_chip: Arc::clone(&range_chip) }
+    }
+
+    /// Convert a vector of AssignedValue to a SafeType. The number of bytes of inputs must equal to the number of bytes of outputs.
+    /// This function also add contraints that a AssignedValue in inputs must be in the range of a byte.
```

Comment:
You said it above, but also add here that inputs are processed little endian?

Reply:
Yes let me update the comment.